### PR TITLE
mbedtls: implement hardware support in mbedtls

### PR DESCRIPTION
**Pull request by an4967**
- net/tls: add see apis for security
- net/tls: add sss hw interface parts in mbedtls 
- net/tls: change HW_ECDSA_VERIFICATION config location
- net/tls: add see_misc file for supporting miscellaneous functions 
- net/tls: apply mbedtls patch to prevent bypass of authentication issue
- sss: change ispdriver library for sss and add header files

### CODE REVIEW

**Comment by davidfather** [comment]
Could you remove the bracket from the title? You may use a colon instead.

**Comment by an4967** [comment]
@davidfather ok, I changed the title.

File: external/mbedtls/x509_crt.c [hunk 1]
```diff
@@ -2103,15 +2104,28 @@ int mbedtls_x509_crt_verify_with_profile(mbedtls_x509_crt *crt, mbedtls_x509_crt
...
+	/* prevent misuse of the vrfy callback - VERIFY_FAILED would be ignored by
+	 * the SSL module for authmode optional, but non-zero return from the
+	 * callback means a fatal error so it shouldn't be ignored
+	 */
+	if (ret == MBEDTLS_ERR_X509_CERT_VERIFY_FAILED) {
+		ret = MBEDTLS_ERR_X509_FATAL_ERROR;
+	}
+
+	if (ret != 0) {
+		*flags = (uint32_t) -1;
```
Comment (sunghan-chang) [inline external/mbedtls/x509_crt.c]: Please remove a space after type casting.
```
(uint32_t) -1; => (uint32_t)-1;
```

File: external/include/mbedtls/see_misc.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
+/****************************************************************************
+ *
+ * Copyright 2016 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline external/include/mbedtls/see_misc.h]: 2017

File: external/mbedtls/see_misc.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
+/****************************************************************************
+ *
+ * Copyright 2016 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline external/mbedtls/see_misc.c]: 2017

File: external/mbedtls/pk_wrap.c [hunk 4]
```diff
@@ -95,15 +95,16 @@ static int rsa_verify_wrap(void *ctx, mbedtls_md_type_t md_alg, const unsigned c
 {
 	int ret;
 
-	if (sig_len < ((mbedtls_rsa_context *)ctx)->len) {
+	if (sig_len < ((mbedtls_rsa_context *) ctx)->len) {
```
Comment (sunghan-chang) [inline external/mbedtls/pk_wrap.c]: revert this line. violation of coding rule

File: external/mbedtls/pk_wrap.c [hunk 5]
```diff
@@ -95,15 +95,16 @@ static int rsa_verify_wrap(void *ctx, mbedtls_md_type_t md_alg, const unsigned c
...
 	}
 
-	if ((ret = mbedtls_rsa_pkcs1_verify((mbedtls_rsa_context *)ctx, NULL, NULL, MBEDTLS_RSA_PUBLIC, md_alg, (unsigned int)hash_len, hash, sig)) != 0) {
+	ret = mbedtls_rsa_pkcs1_verify((mbedtls_rsa_context *) ctx, NULL, NULL, MBEDTLS_RSA_PUBLIC, md_alg, (unsigned int)hash_len, hash, sig);
+	if (ret != 0) {
 		return (ret);
 	}
 
-	if (sig_len > ((mbedtls_rsa_context *)ctx)->len) {
+	if (sig_len > ((mbedtls_rsa_context *) ctx)->len) {
```
Comment (sunghan-chang) [inline external/mbedtls/pk_wrap.c]: same as above

File: external/mbedtls/pk_wrap.c [hunk 6]
```diff
@@ -112,75 +113,108 @@ static int rsa_verify_wrap(void *ctx, mbedtls_md_type_t md_alg, const unsigned c
 
 static int rsa_sign_wrap(void *ctx, mbedtls_md_type_t md_alg, const unsigned char *hash, size_t hash_len, unsigned char *sig, size_t *sig_len, int (*f_rng)(void *, unsigned char *, size_t), void *p_rng)
 {
-	*sig_len = ((mbedtls_rsa_context *)ctx)->len;
+	int ret;
+	
+#if defined(CONFIG_TLS_WITH_SSS) && defined(CONFIG_SUPPORT_FULL_SECURITY)
+	if (!see_check_keyindex(((mbedtls_rsa_context *) ctx)->key_index)) {
```
Comment (sunghan-chang) [inline external/mbedtls/pk_wrap.c]: remove a space between type casting and variable.
Same for every these codes (There are too many)

File: external/include/mbedtls/see_api.h [hunk 7]
```diff
@@ -45,6 +45,11 @@
 #include "../../../os/arch/arm/src/s5j/sss/isp_driver_error.h"
 #include "../../../os/arch/arm/src/s5j/sss/isp_oid.h"
 
+#include "../../../os/arch/arm/src/s5j/sss/isp_driver_aes_securekey.h"
```
Comment (sunghan-chang) [inline external/include/mbedtls/see_api.h]: 1. Why do you add this in here? The mbedtls is a common module, not a architecture specific module.
2. If you want to include these, you had better add that path at makefile first.

File: external/include/mbedtls/see_cert.h [hunk 8]
```diff
@@ -0,0 +1,18 @@
+#define SEE_BEGIN_CERT           "-----BEGIN CERTIFICATE-----\n"
```
Comment (sunghan-chang) [inline external/include/mbedtls/see_cert.h]: Need samsung license

File: external/mbedtls/see_api.c [hunk 9]
```diff
@@ -328,3 +338,1092 @@ int see_check_keyindex(unsigned int index)
...
+	}
+
+	t_buf = malloc(MAX_DATA_SIZE);
+
+	if (t_buf == NULL) {
+		return SEE_ALLOC_ERROR;
+	}
+
+	if (see_mutex_lock(&m_handler) != SEE_OK) {
+		return SEE_MUTEX_LOCK_ERROR;
```
Comment (sunghan-chang) [inline external/mbedtls/see_api.c]: Don't we need to free t_buf in mutex case?

File: external/mbedtls/Kconfig.protocol [hunk 10]
```diff
@@ -5,6 +5,13 @@ config TLS_WITH_SSS
 	depends on S5J_SSS
 	default n
 
+config SUPPORT_FULL_SECURITY
```
Comment (sunghan-chang) [inline external/mbedtls/Kconfig.protocol]: What's the meaning of full security?

**Comment by an4967** [comment]
@sunghan-chang I fixed memory leak bug and applied coding rule

**Pull request merged**